app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/PeerGroupHandshake.kt

```diff
@@ -0,0 +1,111 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.localFileTransfer
+
+import android.net.wifi.p2p.WifiP2pInfo
+import android.util.Log
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.withContext
+import org.kiwix.kiwixmobile.core.BuildConfig
+import java.io.InputStream
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+import java.io.OutputStream
+import java.net.InetAddress
+import java.net.InetSocketAddress
+import java.net.ServerSocket
+import java.net.Socket
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once two peer devices are connected through wifi direct, this task is executed to perform a
+ * handshake between the two devices. The purpose of the handshake is to allow the file sending
+ * device to obtain the IP address of the file receiving device (When the file sending device
+ * is the wifi direct group owner, it doesn't have the IP address of the peer device by default).
+ *
+ * After obtaining the IP address, the sender also shares metadata regarding the file transfer
+ * (no. of files & their names) with the receiver.
+ */
+abstract class PeerGroupHandshake(private var groupInfo: WifiP2pInfo) {
+  private val HANDSHAKE_MESSAGE = "Request Kiwix File Sharing"
+  suspend fun handshake(): InetAddress? =
+    withContext(Dispatchers.IO) {
+      if (BuildConfig.DEBUG) {
+        Log.d(TAG, "Handshake in progress")
+      }
+      if (groupInfo.groupFormed && groupInfo.isGroupOwner && this.isActive) {
+        try {
+          ServerSocket(PEER_HANDSHAKE_PORT)
+            .use { serverSocket ->
+              serverSocket.reuseAddress = true
+              val server = serverSocket.accept()
+              val objectInputStream = ObjectInputStream(server.getInputStream())
+              val kiwixHandShakeMessage = objectInputStream.readObject()
+
+              // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+              return@withContext if (isKiwixHandshake(kiwixHandShakeMessage) && this.isActive) {
+                if (BuildConfig.DEBUG) {
+                  Log.d(TAG, "Client IP address: " + server.inetAddress)
+                }
+                exchangeFileTransferMetadata(server.getInputStream(), server.getOutputStream())
+                server.inetAddress
+              } else {
+                // Selected device is not accepting wifi direct connections through the kiwix app
+                null
+              }
+            }
+        } catch (ex: Exception) {
+          ex.printStackTrace()
+          return@withContext null
```

Comment:
we can get rid of the `return@withContext` by returning the if and adding an else as the final statement. perhaps then it would be nicer as a when eg
```
return when{
groupInfo.groupFormed && groupInfo.isGroupOwner && this.isActive -> readHandshakeAndExchangeMetaData()
...
else -> null
}
```

extracting methods for these large bodies would certainly add a lot of readability to what is happening